### PR TITLE
Show errors in forms in authentication flow

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
- Fix https://github.com/enso-org/cloud-v2/issues/1422
  - Show errors on "login" page by switching to custom Form component
  - Also convert "registration", "reset password" and "forgot password" pages to use the new component
- Preserve email when navigating between auth pages

### Important Notes
None

### Screenshots
After:
image

Before:
image

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] ~~Unit tests have been written where possible.~~

### CODE REVIEW

File: app/dashboard/src/components/AriaComponents/Form/Form.tsx [hunk 1]
```diff
@@ -147,30 +158,13 @@ export const Form = React.forwardRef(function Form<
       register: (name, options) => {
         const registered = register(name, options)
 
-        /**
```
Comment (MrFlashAccount) [inline app/dashboard/src/components/AriaComponents/Form/Form.tsx:150]: Are you sure we can now safely remove it? The mapping is essential to support both native and `react-aria` based components.

File: app/dashboard/src/pages/authentication/Registration.tsx [hunk 2]
```diff
@@ -74,56 +67,50 @@ export default function Registration() {
...
-      footer={
-        <Link to={appUtils.LOGIN_PATH} icon={GoBackIcon} text={getText('alreadyHaveAnAccount')} />
-      }
-      onSubmit={async (event) => {
-        event.preventDefault()
-        setIsSubmitting(true)
-        await auth.signUp(email, password, organizationId)
-        setIsSubmitting(false)
+      footer={<Link to={LOGIN_PATH} icon={GoBackIcon} text={getText('alreadyHaveAnAccount')} />}
+      onSubmit={async ({ email, password, confirmPassword }): Promise<void> => {
```
Comment (MrFlashAccount) [inline app/dashboard/src/pages/authentication/Registration.tsx]: What was the reason to use manual verification logic instead of using `schema` based approach?

**Review by MrFlashAccount — COMMENTED** [review]
Will do the ux review a bit later. Here are some questions I have:

File: app/dashboard/src/components/AriaComponents/Form/Form.tsx (hunk 1 above)
Comment (somebody1234) [reply app/dashboard/src/components/AriaComponents/Form/Form.tsx:150]: it's not being removed, the function was moved to the top and the wrapping was inlined into the `const result`. can be reverted but i just figured it lets us avoid typing the same long generic type three times

File: app/dashboard/src/pages/authentication/Registration.tsx (hunk 2 above)
Comment (somebody1234) [reply app/dashboard/src/pages/authentication/Registration.tsx]: the reason is i forgor about `schema` :skull:

**Comment by somebody1234** [comment]
umm actually the validation errors aren't appearing. the form isnt submittinng at least, so i guess at least it *works*

File: app/dashboard/src/pages/authentication/Login.tsx [hunk 3]
```diff
@@ -33,119 +33,92 @@ import TextLink from '#/components/TextLink'
...
 
+  const signInWithGoogleMutation = useMutation({ mutationFn: auth.signInWithGoogle })
+  const signInWithGitHubMutation = useMutation({ mutationFn: auth.signInWithGitHub })
+  const signInWithPasswordMutation = useMutation({
+    mutationFn: ({ email, password }: { email: string; password: string }) =>
+      auth.signInWithPassword(email, password),
+  })
+  const signInWithGoogle = signInWithGoogleMutation.mutateAsync
+  const signInWithGitHub = signInWithGitHubMutation.mutateAsync
+  const signInWithPassword = signInWithPasswordMutation.mutateAsync
```
Comment (MrFlashAccount) [inline app/dashboard/src/pages/authentication/Login.tsx]: Not sure it's a great idea to use mutations for login. Since we send raw passwords, it might be possible to get that password from the mutation cache later after logging in.

File: app/dashboard/src/pages/authentication/Login.tsx [hunk 4]
```diff
@@ -33,119 +33,92 @@ import TextLink from '#/components/TextLink'
...
-  const [password, setPassword] = React.useState('')
-  const [isSubmitting, setIsSubmitting] = React.useState(false)
-  const shouldReportValidityRef = React.useRef(true)
-  const formRef = React.useRef<HTMLFormElement>(null)
-
-  const localBackend = backendProvider.useLocalBackend()
+  const localBackend = useLocalBackend()
   const supportsOffline = localBackend != null
 
+  const signInWithGoogleMutation = useMutation({ mutationFn: auth.signInWithGoogle })
```
Comment (MrFlashAccount) [inline app/dashboard/src/pages/authentication/Login.tsx]: Not sure if we need additional wrappers here, or we can pull them up to AuthProvider?

File: app/dashboard/src/pages/authentication/Login.tsx (hunk 3 above)
Comment (somebody1234) [reply app/dashboard/src/pages/authentication/Login.tsx]: good point... i think they may (?) have been required before some refactor? but the other pages don't use mutations so no idea why i added them here

File: app/dashboard/src/pages/authentication/Login.tsx (hunk 4 above)
Comment (somebody1234) [reply app/dashboard/src/pages/authentication/Login.tsx]: i mean, do we need mutations at all?
if we do want a mutation for everything then... maybe we could imperatively get the form values to avoid sending them as variables, i guess that's not quite ideal though...

**Comment by MrFlashAccount** [comment]
I pushed some changes in your branch, please take a look(https://github.com/enso-org/enso/pull/10739/commits/4eafaf8df99c437c5116fd40143bb5f27335909c).

TLDR: I fixed the `addonStart` and `addonEnd` visuals, improved the schema, and extracted `password` schema into a separate function to reuse it across the auth pages.

According the UI/UX:

1. Can we bring back the icons for the inputs?
2. Can we create a `Password` component that wraps the basic `Input` but with the ability to toggle the password visibility?
3. Not sure it's a great idea to redirect a user to the `create user` page if the login is invalid. What if the user made a typo? this might also be an attack vector: someone might easily figure out if a user has an account. 
4. Buttons that navigate to the registration/forgot password should autofill the email if the user already typed it on the page, the same goes for `registration/forgot password` pages
5. register/forgot password form inputs should have labels

File: app/dashboard/src/components/AriaComponents/Form/Form.tsx (hunk 1 above)
Comment (MrFlashAccount) [reply app/dashboard/src/components/AriaComponents/Form/Form.tsx:150]: Sorry, missed that 🤦

File: app/dashboard/src/components/AriaComponents/Inputs/Password/Password.tsx [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  TFieldName extends Path<TFieldValues>,
+  TTransformedValues extends FieldValues<Schema> | undefined = undefined,
+>(props: PasswordProps<Schema, TFieldValues, TFieldName, TTransformedValues>) {
+  const [showPassword, setShowPassword] = useState(false)
+
+  return (
+    <Input
+      {...props}
+      type={showPassword ? 'text' : 'password'}
+      addonMiddle={
```
Comment (MrFlashAccount) [inline app/dashboard/src/components/AriaComponents/Inputs/Password/Password.tsx]: Small question: why do we need `addonMiddle`, why neither `addonStart` nor `addonEnd` aren't sufficient here?

File: app/dashboard/src/components/AriaComponents/Inputs/Password/Password.tsx [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        <Button
+          size="medium"
+          variant="icon"
+          icon={showPassword ? EyeIcon : EyeCrossedIcon}
+          className="absolute right-0 top-1/2 -translate-y-1/2"
```
Comment (MrFlashAccount) [inline app/dashboard/src/components/AriaComponents/Inputs/Password/Password.tsx]: `position:absolute` here breaks visuals of the component - input shall not be under the button

CleanShot 2024-08-06 at 18 11 15@2x

File: app/dashboard/e2e/actions/BaseActions.ts [hunk 7]
```diff
@@ -155,4 +155,24 @@ export default class BaseActions implements Promise<void> {
...
+
+  /** Expect an input to have an error (or no error if the expected value is `null`).
+   * If the expected value is `undefined`, the assertion is skipped. */
+  expectInputError(name: string, description: string, expected: string | null | undefined) {
+    if (expected === undefined) {
+      return this
+    } else if (expected != null) {
+      return this.step(`Expect ${description} error to be '${expected}'`, async (page) => {
+        // eslint-disable-next-line no-restricted-properties
+        const element = page.locator(`fieldset:has(input[name=${name}]) > span.text-danger`)
```
Comment (MrFlashAccount) [inline app/dashboard/e2e/actions/BaseActions.ts]: This is a fragile selector. Can we use `testid` or something like that here?

File: app/dashboard/src/components/AriaComponents/Inputs/Password/Password.tsx (hunk 5 above)
Comment (somebody1234) [reply app/dashboard/src/components/AriaComponents/Inputs/Password/Password.tsx]: a) we need the eye to be centered on the text input. maybe we always know the height so we can position using `rem` without having to use `top-1/2`?
b) i was working under the assumption that users of `<Password` may want to add their own `addonStart`/`addonEnd`, although it's true that doing this still prevents them from adding their own `addonMiddle`

File: app/dashboard/src/components/AriaComponents/Inputs/Password/Password.tsx (hunk 6 above)
Comment (somebody1234) [reply app/dashboard/src/components/AriaComponents/Inputs/Password/Password.tsx]: true... but then i think the best way would be to do it like the old component and add the eye as an option to `Input` directly so that we get full control over styling/markup structure

File: app/dashboard/e2e/actions/BaseActions.ts (hunk 7 above)
Comment (somebody1234) [reply app/dashboard/e2e/actions/BaseActions.ts]: the entire thing? or the `.text-danger` specifically?
i'm thinking of keeping the `fieldset:has()` since i don't think *that* part is too fragile, but do agree that the `text-danger` isn't great

File: app/dashboard/src/components/AriaComponents/Inputs/Password/Password.tsx (hunk 5 above)
Comment (MrFlashAccount) [reply app/dashboard/src/components/AriaComponents/Inputs/Password/Password.tsx]: A. Looks like the naming "addonStart" and "addonEnd" is misleading 🥲 start and end here means "left" and "right", but they _might_ flip based on the locale(eg for RTL languages). Maybe there's a better naming for that?

B. we can use React.Fragment and stack multiple addons

File: app/dashboard/e2e/actions/BaseActions.ts (hunk 7 above)
Comment (MrFlashAccount) [reply app/dashboard/e2e/actions/BaseActions.ts]: Both 'fieldset', '>' and 'text-danger' are fragile parts here

File: app/dashboard/src/components/AriaComponents/Inputs/Password/Password.tsx (hunk 6 above)
Comment (MrFlashAccount) [reply app/dashboard/src/components/AriaComponents/Inputs/Password/Password.tsx]: I guess using addonEnd should be sufficient?

File: app/dashboard/src/components/AriaComponents/Inputs/Password/Password.tsx (hunk 5 above)
Comment (somebody1234) [reply app/dashboard/src/components/AriaComponents/Inputs/Password/Password.tsx]: oh... yeah, makes sense. i'll mark as resolved since this is addressed locally but open an issue since i do think the naming is a bit confusing

File: app/dashboard/e2e/actions/BaseActions.ts (hunk 7 above)
Comment (somebody1234) [reply app/dashboard/e2e/actions/BaseActions.ts]: adjusted, but not sure whether the current solution is ideal. the label containing the input is causing a bit of awkwardness (we can't use exact locators)

File: app/dashboard/src/components/AriaComponents/Inputs/variants.ts [hunk 8]
```diff
@@ -41,11 +41,17 @@ export const INPUT_STYLES = tv({
         base: 'border-[0.5px] outline-offset-2 border-primary/20 focus-within:outline focus-within:outline-2 focus-within:outline-offset-[-1px] focus-within:outline-primary focus-within:border-primary/50',
         textArea: 'border-transparent focus-within:border-transparent',
       },
+      settings: {
```
Comment (MrFlashAccount) [inline app/dashboard/src/components/AriaComponents/Inputs/variants.ts]: variants shall not be place-specific. 
If we want to override visuals, there are 2 ways we can achieve that:
1. Use `className` prop
2. Pass `tv` schema that extends the default schema. We currently don't have that ability because I don't know how to achieve type-safety here. It'd be awesome if we fix that

File: app/dashboard/e2e/actions/BaseActions.ts (hunk 7 above)
Comment (MrFlashAccount) [reply app/dashboard/e2e/actions/BaseActions.ts]: WDYM? Why we can't use something like: 

```ts
const error = page.getByTestId('password-input').getByTestId('error-text')
```

Comment (somebody1234) [reply app/dashboard/e2e/actions/BaseActions.ts]: i guess the idea is that test-id should be a last resort, but i guess it might be our best option here?

Comment (MrFlashAccount) [reply app/dashboard/e2e/actions/BaseActions.ts]: Yeah, I don't see better options either

**Comment by PabloBuchu** [comment]
On registration form I should be able to move from input to input with tab in order email -> password -> confirm password. Right now when I want to jump from password to confirm I need to jump password -> reveal password icon -> confirm password. 

Also why github and google icons went grey? 

the rest is fine

**Comment by somebody1234** [comment]
> Also why github and google icons went grey?

we switched to using regular buttons so we are using the default colors, i'll see whether we have full color variants maybe?

**Comment by somebody1234** [comment]
> I should be able to move from input to input with tab in order email -> password -> confirm password

unfortunately idk if we can change this, we want those buttons to still be possible to navigate to for accessibility, and cmd+tab, shift+tab and cmd+shift+tab are all taken by system shortcuts

that said, enter technically does work, it's a bit janky though because it works through validation, so you do end up at the right input but it becomes red

**Review by PabloBuchu — APPROVED** [review]

**Comment by PabloBuchu** [comment]
@somebody1234 LGTM good to go
